### PR TITLE
Add an option to disable purge in delete room admin API

### PR DESCRIPTION
**Pull request by dklimpel**
Add option ```purge``` to ```POST /_synapse/admin/v1/rooms/<room_id>/delete```
Fixes: #3761

So you can remove rooms without repointing users elsewhere an do not delete from server.
You can shutdown a room and you can decide whether you want to delete it or not from your server.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: docs/admin_api/rooms.md [hunk 1]
```diff
@@ -369,7 +369,7 @@ to the new room will have power level `-10` by default, and thus be unable to sp
 If `block` is `True` it prevents new joins to the old room.
 
 This API will remove all trace of the old room from your database after removing
-all local users.
+all local users. If you do not like to remove the trace set `purge` to `False`.
```
Comment (anoadragon453) [inline docs/admin_api/rooms.md:372]: I would probably rewrite this sentence as something like:

> If `purge` is `true` (the default), all traces of the old room will be removed from your database after removing all local users. If you do not want this to happen, set `purge` to `false`.

File: docs/admin_api/rooms.md [hunk 2]
```diff
@@ -430,8 +431,10 @@ The following JSON body parameters are available:
               `new_room_user_id` in the new room. Ideally this will clearly convey why the
                original room was shut down. Defaults to `Sharing illegal content on this server
                is not permitted and rooms in violation will be blocked.`
-* `block` - Optional. If set to `true`, this room will be added to a blocking list, preventing future attempts to
-  join the room. Defaults to `false`.
+* `block` - Optional. If set to `true`, this room will be added to a blocking list, preventing
+            future attempts to join the room. Defaults to `false`.
+* `purge` - Optional. If set to `true`, it will remove all trace of a room from your database.
```
Comment (anoadragon453) [inline docs/admin_api/rooms.md:436]:
```suggestion
* `purge` - Optional. If set to `true`, it will remove all traces of the room from your database.
```

File: tests/rest/admin/test_room.py [hunk 3]
```diff
@@ -351,6 +368,41 @@ def test_purge_room_and_not_block(self):
         self._is_blocked(self.room_id, expect=False)
         self._has_no_members(self.room_id)
 
+    def test_block_room_and_not_purge(self):
+        """Test to block a room and do not purge it.
```
Comment (anoadragon453) [inline tests/rest/admin/test_room.py:372]:
```suggestion
        """Test to block a room without purging it.
```

File: tests/rest/admin/test_room.py [hunk 4]
```diff
@@ -351,6 +368,41 @@ def test_purge_room_and_not_block(self):
...
+        Members will not be moved to a new room and will not receive a message.
```
Comment (anoadragon453) [inline tests/rest/admin/test_room.py:373]:
```suggestion
        Members will not be moved to a new room and will not receive a message.
        The room will not be purged.
```

File: tests/rest/admin/test_room.py [hunk 5]
```diff
@@ -351,6 +368,42 @@ def test_purge_room_and_not_block(self):
...
+        with self.assertRaises(AssertionError):
+            self._is_purged(self.room_id)
+
+        # Test that room is not blocked
+        self._is_blocked(self.room_id, expect=False)
+
+        # Assert one user in room
+        self._is_member(room_id=self.room_id, user_id=self.other_user)
+
+        body = json.dumps({"block": False, "purge": False})
```
Comment (DMRobertson) [inline tests/rest/admin/test_room.py:386]: Is this a typo? Should this have been `"block": True`?

Comment (dklimpel) [reply tests/rest/admin/test_room.py:386]: Good catch. I think so.
Neither block nor purge is certainly not a good use case.
My PR #11223 needs an update, too.

Comment (DMRobertson) [reply tests/rest/admin/test_room.py:386]: > Neither block nor purge is certainly not a good use case.

To check my understanding, I think no block and no purge would
- kick all local users out of the room
- retain events for that room
- all users would be free rejoin

which I agree, sounds like it ought to caught and rejected with 400 bad request

Comment (dklimpel) [reply tests/rest/admin/test_room.py:386]: I don't think you have to check everything. Maybe it's a edge case.

Comment (DMRobertson) [reply tests/rest/admin/test_room.py:386]: My concern is that someone might accidentally use the API in this way, and not realise that the room can come back to haunt them.

(Is there a use case for shutting down the room with `block: False` at all?)

Comment (dklimpel) [reply tests/rest/admin/test_room.py:386]: I have no idea for an use case `{"block": False, "purge": False}` and remove all users.
